ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -57,80 +55,112 @@ import           Ouroboros.Network.Mux.Types
 --
 
 -- |
--- Specification for peers of a protocol.  This type instructs the multiplexing
--- layer to run a given client \/ server peers.
+-- Application run by mux layer.
+--
+-- * enumeration of client application, e.g. a wallet application communicating
+--   with a node using ChainSync and TxSubmission protocols; this only requires
+--   to run client side of each protocol.
+--
+-- * enumeration of server applications: this application type is mostly useful
+--   tests.
 --
-data MuxPeer m where
-  -- |
-  -- A non pipeliend peer together with a codec.
-  --
-  OnlyClient
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsClient st m a
-    -> MuxPeer m
-
-  -- |
-  -- A pipelined peer together with a codec.
-  --
-  OnlyPipelinedClient
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> MuxPeer m
-
-  -- |
-  -- Server peer with a codec
-  OnlyServer
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Client and server peers with the corresponding codec.
-  --
-  ClientAndServer
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Pipelined client and a server with the correspnding codec.
-  --
-  PipelinedClientAndServer
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
+-- * enumeration of both client and server applications, e.g. a full node
+--   serving downstream peers using server side of each protocol and getting
+--   updates from upstream peers using client side of each of the protocols.
+--
+data MuxApplication ptcl m where
+  MuxClientApplication
+    -- Client application; most simple application will be @'runPeer'@ or
+    -- @'runPipelinedPeer'@ supplied with a codec and a @'Peer'@ for each
+    -- @ptcl@.  But it allows to handle resources if just application of
+    -- @'runPeer'@ is not enough.  It will be run as @'muxInitiator'@.
+    :: (ptcl -> Channel m ByteString ->  m ())
```

Comment:
Indeed, when a client connects to some upstream pear, the mux layer will run all the mini-protocols.  This might be limiting, but I think that's actually a pros, e.g. when one wants to connect to a node with just a single `ChanSync` mini-protocol and one knows that the node runs it among it's mini protocols on a given port it will be impossible to connect to that port; the server has to run the same set of protocol.  Imagine  a core node, that doesn't want to speak with non core nodes.